feat: allow overriding of protocol methods in subclasses

Sinclert opened the pull request:
This PR addresses issue https://github.com/scikit-hep/vector/issues/127 by extending the [_handler_of](https://github.com/scikit-hep/vector/blob/5ba111555bae20887e417c1e73101a940d06fc57/src/vector/_methods.py#L2536-L2556) function to consider the module of all the inherited classes of a certain object. This is neccesary to allow the definition of client-side custom classes that can inherit from the library ones, and override their protocol methods (`__iadd__`, `__isub__` ...).

### Additional clarifications
- I inverted the logic of the `if isinstance(obj, Vector)` to reduce indentation.
- I imported Python's `contextlib.suppress` helper to reduce indentation.

cc @jpivarski

jpivarski commented:
I personally think that indented control structures are more clear than imperative-jumping logic like `break` and `continue`, but I'm fine with it being written like this. I checked on [contextlib.suppress](https://docs.python.org/3/library/contextlib.html#contextlib.suppress) and it was introduced in Python 3.4, but Vector is a Python 3-only library, so that's fine.

~~Now, thinking about the substance of this: what this change does is it makes instances of any subclasses of VectorObject, VectorNumpy, or VectorAwkward be "lower priority" than any of these known types. That means that when a CustomVector is used in an expression with any other kind of vector, that other vector is used to compute it, regardless of whether CustomVector is derived from VectorObject, VectorNumpy, or VectorAwkward. That's probably not what you want. The reason we have these priority rules is so that `single_vector_object + array_of_vectors` gets computed by the `array_of_vectors`'s methods, which broadcasts the `single_vector_object` to every item in the array. If it's an array-of-vectors class that gets inherited, you'd want that derived class to win over VectorObject, so that the VectorObject gets broadcasted.~~

~~What you originally described in issue #127 would be better: walk the `.mro` of the class object, checking to see if its `__module__` is in `_handler_priority`, and stop when you find it, returning that as its priority. That way, a CustomSingleVector inheriting from VectorObject would have lower priority than CustomArrayOfVectors inheriting from VectorNumpy or VectorAwkward, and you can be sure to get the right broadcasting behavior from the array classes.~~

Nevermind: that's what you're doing already. I was misled by the "return -1 if not handler is found"—that can't happen because `_get_handler_index` is called on instances of `VectorProtocol`; they _must_ have a VectorObject, VectorNumpy, or VectorAwkward (or other, future backends) in their inheritance tree by construction, so you could `assert index != -1` before returning. Objects that aren't `Vector`s are `continue`d before `_get_handler_index` can be called on them.

So if you think this PR is ready to merge, I am. (I'd only change the docstring of `_get_handler_index` to not mention the `-1` case and maybe add an assertion for it.)

jpivarski commented:
Thanks!

Sinclert commented:
You are completely right: if the `_get_handler_index` function is only called within the `_handler_of` function for-loop, then all objects are going to be a `Vector` or some sort.

I addressed your comment on https://github.com/scikit-hep/vector/pull/128/commits/9c60f5fa5e724a7b59b2fdaf8d5fe24acc18b1ed, and I think it is ready to merge.

henryiii commented on src/vector/_methods.py:
```diff
@@ -2533,6 +2534,18 @@ def _from_signature(
 
+def _get_handler_index(obj: VectorProtocol) -> int:
+    """Returns the index of the first valid handler checking the list of parent classes"""
+    index = -1
+    for cls in type(obj).__mro__:
+        with suppress(ValueError):
+            index = _handler_priority.index(cls.__module__)
+            break
+
+    assert index != -1
+    return index
```
```suggestion
    for cls in type(obj).__mro__:
        with suppress(ValueError):
            return _handler_priority.index(cls.__module__)
    raise IndexError(f"Could not find a valid handler for {obj}! This should not happen.")

    assert index != -1
    return index
```

jpivarski replied:
Shouldn't it be an AssertionError, since it would be an internal logic error, not a user error?

Otherwise, this implementation is cleaner (no need to introduce the `-1`).

henryiii replied:
Probably, yes. It should not be an assert, as the main reason to use assert is to have the ability to turn off the assertions, but this one should never be reached, and we do always want to throw if we reach it.

henryiii replied:
(Changed in the suggestion)

jpivarski replied:
But it's also true that an assertion represents something that should never happen. If there's ever a situation in which an assertion would fail, wrong behavior would result from letting it slip by. Turning off assertions trades safety for speed—which can be a reasonable thing to do when you're confident—but the above argument would apply to every assertion, not just this one, and then what would assertions be for?

(Note, I'm saying that it shouldn't be possible for this to get out of the `for` loop due to constraints applied elsewhere, not that it is possible and we just shouldn't allow it. That's why I think it's an assertion.)

If you're happy with it, I'll commit this suggestion.

Sinclert replied:
I am OK with it too. Feel free to commit.

henryiii replied:
If something is never reached, turning it off doesn't speed anything up, so there's no need to turn it off. Asserts should be optional checks, not fallbacks when nothing else works. That's why the assert statement takes a condition - it's a check, not a fallback.

Unrelated to this PR, but slightly to the general discussion, Python 3.11 will finally make a `try` block virtually no-cost if there is no throw. Currently it's less expensive than an `if` if there is no throw, but non-zero.

henryiii replied:
Actually, this is not the best way to do this, as handling the exception is expensive.

jpivarski replied:
There wouldn't be a speed difference here, since (in principle) the `raise AssertionError` line is never reached. I was referring to not having to compute expressions "`xyz`" in an `assert xyz` statement in general—the only reason I can imagine for turning off assertions.

henryiii commented on src/vector/_methods.py:
```diff
@@ -2544,13 +2557,12 @@ def _handler_of(*objects: VectorProtocol) -> VectorProtocol:
-                type(obj).__module__
-            ) > _handler_priority.index(type(handler).__module__):
-                handler = obj
+        if not isinstance(obj, Vector):
+            continue
```
This is "guard" style, and I have come to like it. While the continue is slightly irritating, it simplifies the logic for the reader. This is "guarded", that is, you read "do nothing if this is not a vector", while the nested block before requires you look to see there's no else block, and requires more layers (mentally, as well as visually). If you have even more deeply nested structures, this is a must.

I am less fond of the guard inside a loop as I would be elsewhere, but I do think it's an improvement.

henryiii replied:
(The "normal" guard clause would be `if cond: return` at the top of a function)

jpivarski replied:
I frequently use guard style when the error case raises an exception or returns early from a function.

But personally, I find any logic involving `break` and `continue` to be less clear than fenced-off blocks (functional programming influence), though I also don't have strong enough feelings to think it ought to be changed. If you (both) think it's easier to read, that's reason enough to keep it.

henryiii replied:
Those are the best places to use it. :) I think it's very close here, and I'd not probably ask for one over the other in a review, but seeing both, I do have a very minor preference for the guard.

"break" specifically is very useful in Python loops due to the else clause.

Sinclert commented:
Just checking: would these changes produce a new minor version release? or you guys have other time-frame in mind?

I am asking because the current MadMiner approach to the subclassing is broken, and I would like to know if we should create a temporal solution, or just wait for... -check notes- `vector 0.8.4` to be released.

jpivarski commented:
I think we can push a new patch release as soon as this is merged.
